File: Dynamic Programming/Palindrome-Partitioning-131.cpp
// Given a string s, partition s such that every substring of the partition is a palindrome. Return all possible palindrome partitioning of s.
// A palindrome string is a string that reads the same backward as forward.

// Example 1:

// Input: s = "aab"
// Output: [["a","a","b"],["aa","b"]]
// Example 2:

// Input: s = "a"
// Output: [["a"]]

// Constraints:

// 1 <= s.length <= 16
// s contains only lowercase English letters.

class Solution {
public:
    vector<vector<string>> res;
    bool isPalindrome(string s){
        int i=0, j = s.size()-1;
        while(i<=j){
            if(s[i++]!=s[j--])
                return false;
        }
        return true;
    }
    
    void solve(string s, int index, vector<string> &path){
        if(index == s.size()){
            res.push_back(path);
            return;
        }
        for(int i=index; i<s.size(); ++i){
            string temp = s.substr(index, i-index+1);
            if(isPalindrome(temp)){
                path.push_back(temp);
                solve(s, i+1, path);
                path.pop_back(); 
            }           
        }
    }
    
    vector<vector<string>> partition(string s) {
        vector<string> path;
        solve(s, 0, path);
        return res;
    }
};

File: Backtracking/PalindromePartitioning131.java
// Given a string s, partition s such that every substring of the partition is a palindrome. 
// Return all possible palindrome partitioning of s.
// A palindrome string is a string that reads the same backward as forward. 

// Example 1:
// Input: s = "aab"
// Output: [["a","a","b"],["aa","b"]]
// Example 2:
// Input: s = "a"
// Output: [["a"]]
 
// Constraints:
// 1 <= s.length <= 16
// s contains only lowercase English letters.

class Solution {
    List<List<String>> ans = new ArrayList<>();
    
    public void solve(String s, List<String> path, int start){
        if(start >= s.length()){
            ans.add(new ArrayList(path));
            return;
        }
        for(int i = start; i < s.length(); i++){
            String str = s.substring(start, i + 1);
            if(isPalindrome(str)){
                path.add(str);
                solve(s, path, i+1);
                path.remove(path.size() - 1);
            }
        }
    }
    
    public List<List<String>> partition(String s) {
        solve(s, new ArrayList<>(), 0);
        return ans;
    }
    
    public boolean isPalindrome(String s){
        int start = 0, end = s.length() - 1;
        while(start <= end){
            if(s.charAt(start) != s.charAt(end))
                return false;
            start++;
            end--;
        }
        return true;
    }
}